src/ds1302.c

```diff
@@ -2,130 +2,122 @@
         ds_writebyte( DS_CMD_RAM >> 1 | 0x00, MAGIC_LO);
         ds_writebyte( DS_CMD_RAM >> 1 | 0x01, MAGIC_HI);
 
-        //for (i=0; i<4; i++)
-            //ds_writebyte( DS_CMD_RAM >> 1 | (i+2), 0x00);
-	ds_ram_config_write(config);
+	ds_ram_config_write();
 	return;
     }
     
     // read ram config
-    for (i=0; i<4; i++)
-        config[i] = ds_readbyte(DS_CMD_RAM >> 1 | (i+2));
+    for (i=0; i!=4; i++)
+        config_table[i] = ds_readbyte(DS_CMD_RAM >> 1 | (i+2));
 }
 
-void ds_ram_config_write(uint8_t config[4]) {
+void ds_ram_config_write() {
     uint8_t i;
-    for (i=0; i<4; i++)
-        ds_writebyte( DS_CMD_RAM >> 1 | (i+2), config[i]);
-}
-
-void ds_writebit(__bit b) {
-    _nop_; _nop_;
-    DS_IO = b;
-    DS_SCLK = 1;
-    _nop_; _nop_;
-    DS_SCLK = 0;
-}
-
-__bit ds_readbit() {
-    __bit b;
-    _nop_; _nop_;
-    b = DS_IO;
-    DS_SCLK = 1;
-    _nop_; _nop_;
-    DS_SCLK = 0;
-    return b;
+    for (i=0; i!=4; i++)
+        ds_writebyte( DS_CMD_RAM >> 1 | (i+2), config_table[i]);
+}
+
+void sendbyte(uint8_t b)
+{
+  b;
+  __asm
+	push	ar7
+        mov     a,dpl
+	mov	r7,#8
+00001$:
+	nop
+	nop
+        rrc     a
+        mov     _P1_1,c
+	setb	_P1_2
+	nop
+	nop
+	clr	_P1_2
+	djnz	r7,00001$
+	pop	ar7
+  __endasm;
+}
+
+uint8_t readbyte()
+{
+  __asm
+	push	ar7
+	mov 	a,#0
+	mov 	r7,#8
+00002$:
+	nop
+	nop
+	mov	c,_P1_1
+	rrc	a	
+	setb	_P1_2
+	nop
+	nop
+	clr	_P1_2
+	djnz	r7,00002$
+	mov	dpl,a
+	pop	ar7
+  __endasm;
 }
 
 uint8_t ds_readbyte(uint8_t addr) {
     // ds1302 single-byte read
-    uint8_t i, b;
+    uint8_t b;
     b = DS_CMD | DS_CMD_CLOCK | addr << 1 | DS_CMD_READ;
     DS_CE = 0;
     DS_SCLK = 0;
     DS_CE = 1;
     // send cmd byte
-    for (i=0; i < 8; i++) {
-        ds_writebit(b & 0x01);
-        _nop_; _nop_;
-        b>>=1;
-    }
+    sendbyte(b);
     // read byte
-    b=0;
-    for (i=0; i < 8; i++) {
-	b>>=1;
-        if (ds_readbit()) 
-            b |= 0x80; // set
-        _nop_; _nop_;              
-    }
+    b=readbyte();
     DS_CE = 0;
     return b;
 }
 
-void ds_readburst(uint8_t time[8]) {
+void ds_readburst() {
     // ds1302 burst-read 8 bytes into struct
-    uint8_t i, j, b;
+    uint8_t j, b;
     b = DS_CMD | DS_CMD_CLOCK | DS_BURST_MODE << 1 | DS_CMD_READ;
     DS_CE = 0;
     DS_SCLK = 0;
     DS_CE = 1;
     // send cmd byte
-    for (i=0; i < 8; i++) {
-        ds_writebit(b & 0x01);
-        _nop_; _nop_;
-	b>>=1;
-    }
+    sendbyte(b);
     // read bytes
-    for (j=0; j < 8; j++) {
-	b=0;
-        for (i=0; i < 8; i++) {
-	    b>>=1;
-            if (ds_readbit()) 
-                b |= 0x80; // set
-            _nop_; _nop_;              
-        }
-        time[j] = b;
-    }
+    for (j=0; j!=8; j++) 
+        rtc_table[j] = readbyte();
     DS_CE = 0;
 }
 
-uint8_t ds_writebyte(uint8_t addr, uint8_t data) {
+void ds_writebyte(uint8_t addr, uint8_t data) {
     // ds1302 single-byte write
-    uint8_t i, b = 0;
+    uint8_t b = 0;
```

Comment:
this is sort of a side discussion:
I have always been taught, even in assembly that register/variable initialization (zeroing, etc) is always a good thing and a responsible practice.
However, I noticed that my variable initialization and assignment to zero creates the extra instructions in sdcc to assign zero value to the variable.
In several places in main, I already removed the zero assignment (and just have initialization). Perhaps this is working because it's static ram (and not stack variable/register). I'm curious is this is an optimization (to save one whole instruction ;) that can be done in functions?